### PR TITLE
add zmap io and no_data param in write_asc

### LINKED ISSUE #197: BUG: write_asc_grid function does not run as expected
### kriging_tools.py  line 73 
I got  abs((x[-1] - x[0]) / (x.shape[0] - 1))   calculated as 0.999999999999998 , while dx   calculated as 1, so the function didn't work.
Is there any way to overcome this problem

### LINKED ISSUE #198: zmap IO
Good evening,

I have used PyKrige for our research at TNO, which is really helpful!
I have added .zmap (useful for Petrel and QGIS) reading and writing functionality to kriging_tools.py. Would you be interested in adding this? 
I tried to push my branch but did not get permission to do so.

Kind regards,
Mark Vrijlandt
TNO

### PR DESCRIPTION
**Pull request by MarkTNO**
Example zmap:
https://saurabhkukade.com/posts/2020/07/understanding-zmap-file-format/

### CODE REVIEW

File: pykrige/kriging_tools.py [hunk 1]
```diff
@@ -71,8 +75,8 @@ def write_asc_grid(x, y, z, filename="output.asc", style=1):
     dx = abs(x[1] - x[0])
     dy = abs(y[1] - y[0])
     if (
-        abs((x[-1] - x[0]) / (x.shape[0] - 1)) != dx
-        or abs((y[-1] - y[0]) / (y.shape[0] - 1)) != dy
+            abs((x[-1] - x[0]) / (x.shape[0] - 1)) != dx
+            or abs((y[-1] - y[0]) / (y.shape[0] - 1)) != dy
     ):
         raise ValueError(
             "X or Y spacing is not constant; *.asc grid cannot be written."
```
Comment (MuellerSeb) [inline pykrige/kriging_tools.py]: Could you replace these float comparisons with:
```python
    if (
                    not np.isclose(abs((x[-1] - x[0]) / (x.shape[0] - 1)), dx)
                    or not np.isclose(abs((y[-1] - y[0]) / (y.shape[0] - 1)), dy)
            ):
                raise ValueError(
                    "X or Y spacing is not constant; *.asc grid cannot be written."
                )
```
to address #197?

File: pykrige/kriging_tools.py [hunk 2]
```diff
@@ -246,3 +248,181 @@ def read_asc_grid(filename, footer=0):
...
+        z = np.array(z.tolist(no_data))
+
+    x = np.squeeze(np.array(x))
+    y = np.squeeze(np.array(y))
+    z = np.squeeze(np.array(z))
+    nx = len(x)
+    ny = len(y)
+
+    dx = abs(x[1] - x[0])
+    dy = abs(y[1] - y[0])
```
Comment (MuellerSeb) [inline pykrige/kriging_tools.py:289]: Could you add the same check like in the asc routine for checking the grid?
```python
    if (
                    not np.isclose(abs((x[-1] - x[0]) / (x.shape[0] - 1)), dx)
                    or not np.isclose(abs((y[-1] - y[0]) / (y.shape[0] - 1)), dy)
            ):
                raise ValueError(
                    "X or Y spacing is not constant; *.zmap grid cannot be written."
                )
```
to address #197?

File: pykrige/kriging_tools.py [hunk 3]
```diff
@@ -246,3 +248,181 @@ def read_asc_grid(filename, footer=0):
...
+        for n in range(z.shape[1]):
+            count = 0
+            for m in range(z.shape[0] - 1, -1, -1):
+                count += 1
+                if np.isnan(z[m, n]):
+                    f.write(space_back_to_front(format(no_data, "13.7E") + '  '))
+                else:
+                    if abs(z[m, n]) >= 1E100:
+                        f.write(space_back_to_front(format(z[m, n], "13.7E") + ' '))
+                    elif abs(z[m, n]) >= 1E6:
```
Comment (MuellerSeb) [inline pykrige/kriging_tools.py]: Where do these tolerance values come from?

File: pykrige/kriging_tools.py [hunk 4]
```diff
@@ -246,3 +248,181 @@ def read_asc_grid(filename, footer=0):
...
+    x : numpy array, shape (N,)
+        1D array of N X-coordinates.
+    y : numpy array, shape (M,)
+        1D array of M Y-coordinates.
+    CELLSIZE : tuple or float
+        Either a two-tuple of (x-cell size, y-cell size),
+        or a float that specifies the uniform cell size.
+    NODATA : float
+        Value that specifies which entries are not actual data.
+    COORD_SYS : String
```
Comment (MuellerSeb) [inline pykrige/kriging_tools.py]: could you rename these returns to lower-case names?

**Review by MuellerSeb — CHANGES_REQUESTED** [review]
Thank you for this contribution! 

Some requests from my side:
- I'd love to see the bug mentioned in #197 fixed with this contribution (since the same would apply to your additions)
- see review comments
- please run `black .` in the root directory to format the code accordingly (CI is failing there)

Edit:
- please add a test like for the asc routines:
  https://github.com/GeoStat-Framework/PyKrige/blob/09fe0183aa8e47ceb585a67408e08a4a4ffaced9/tests/test_core.py#L1127

File: pykrige/kriging_tools.py (hunk 3 above)
Comment (MarkTNO) [reply pykrige/kriging_tools.py]: To correct for different string length due to extra exponential digit.

**Review by MuellerSeb — APPROVED** [review]
@MarkTNO thanks for the additions. I applied `black` and pushed to your branch (don't forget to pull if you want to add further changes).

If you think, this is ready, let me know. Then I am happy to merge.

**Comment by MarkTNO** [comment]
This looks good. Thanks.

**Pull request merged**